packages/datetime/src/dateInput.tsx

```diff
@@ -320,43 +311,34 @@ export class DateInput extends AbstractPureComponent<IDateInputProps, IDateInput
         const isInputFocused = didSubmitWithEnter ? true : false;
 
         if (this.props.value === undefined) {
-            this.setState({ isInputFocused, isOpen, value: momentDate, valueString: this.getDateString(momentDate) });
+            this.setState({ isInputFocused, isOpen, value: newDate, valueString: this.getDateString(newDate) });
         } else {
             this.setState({ isInputFocused, isOpen });
         }
-        Utils.safeInvoke(this.props.onChange, date === null ? null : fromMomentToDate(momentDate));
+        Utils.safeInvoke(this.props.onChange, newDate);
     };
 
-    private shouldCheckForDateChanges(prevMomentDate: moment.Moment, nextMomentDate: moment.Moment) {
-        return nextMomentDate != null && !isMomentNull(prevMomentDate) && prevMomentDate.isValid();
+    private shouldCheckForDateChanges(prevDate: Date, nextDate: Date) {
+        return nextDate != null && prevDate != null && isValid(prevDate);
     }
 
-    private hasMonthChanged(prevMomentDate: moment.Moment, nextMomentDate: moment.Moment) {
-        return (
-            this.shouldCheckForDateChanges(prevMomentDate, nextMomentDate) &&
-            nextMomentDate.month() !== prevMomentDate.month()
-        );
+    private hasMonthChanged(prevDate: Date, nextDate: Date) {
+        return this.shouldCheckForDateChanges(prevDate, nextDate) && nextDate.getMonth() !== prevDate.getMonth();
     }
 
-    private hasTimeChanged(prevMomentDate: moment.Moment, nextMomentDate: moment.Moment) {
+    private hasTimeChanged(prevDate: Date, nextDate: Date) {
         return (
-            this.shouldCheckForDateChanges(prevMomentDate, nextMomentDate) &&
+            this.shouldCheckForDateChanges(prevDate, nextDate) &&
             this.props.timePrecision != null &&
-            (nextMomentDate.hours() !== prevMomentDate.hours() ||
-                nextMomentDate.minutes() !== prevMomentDate.minutes() ||
-                nextMomentDate.seconds() !== prevMomentDate.seconds() ||
-                nextMomentDate.milliseconds() !== prevMomentDate.milliseconds())
+            (nextDate.getHours() !== prevDate.getHours() ||
+                nextDate.getMinutes() !== prevDate.getMinutes() ||
+                nextDate.getSeconds() !== prevDate.getSeconds() ||
+                nextDate.getMilliseconds() !== prevDate.getMilliseconds())
         );
     }
 
     private handleInputFocus = (e: React.FocusEvent<HTMLInputElement>) => {
-        let valueString: string;
-        if (isMomentNull(this.state.value)) {
-            valueString = "";
-        } else {
-            valueString = momentToString(this.state.value, this.props.format, this.props.locale);
-        }
-
+        const valueString = this.getDateString(this.state.value);
```

Comment:
The logic here isn't quite the same as `getDateString`, so we can't replace it wholesale. e.g. we don't want to show out-of-range errors or invalid-date error messages here; those should only appear on blur. Instead, we want to show the original, invalid date in the field on focus.

Are there tests checking this behavior?

Reply:
Changed this to use `dateToString` value which shows the original, out of bounds date on focus. But this doesn't handle the invalid dates. Invalid string inputs are attempted to be converted to a date object, which is saved in props. When a string is an invalid date string the prop saved is `Invalid Date` and the string is lost. We can try and change this but it could require us saving the input as a string as well.

Reply:
Changed this to use `dateToString` value which shows the original, out of bounds date on focus. But this doesn't handle the invalid dates. Invalid string inputs are attempted to be converted to a date object, which is saved in props. When a string is an invalid date string the prop saved is `Invalid Date` and the string is lost. We can try and change this but it could require us saving the input as a string as well.